libfreerdp/codec/progressive.c

```diff
@@ -1712,24 +1752,62 @@ static INLINE int progressive_process_tiles(PROGRESSIVE_CONTEXT* progressive, wS
 		return -1044;
 	}
 
+	if (progressive->rfx_context->priv->UseThreads)
+	{
+		work_objects = (PTP_WORK*)calloc(region->numTiles, sizeof(PTP_WORK));
+		if (!work_objects)
+			return -1;
+
+		params = (PROGRESSIVE_TILE_PROCESS_WORK_PARAM*)calloc(
+		    region->numTiles, sizeof(PROGRESSIVE_TILE_PROCESS_WORK_PARAM));
+		if (!params)
+		{
+			free(work_objects);
+			return -1;
+		}
+	}
+
 	for (index = 0; index < region->numTiles; index++)
 	{
 		RFX_PROGRESSIVE_TILE* tile = region->tiles[index];
 
-		switch (tile->blockType)
+		if (progressive->rfx_context->priv->UseThreads)
 		{
-			case PROGRESSIVE_WBT_TILE_SIMPLE:
-			case PROGRESSIVE_WBT_TILE_FIRST:
-				status = progressive_decompress_tile_first(progressive, tile, region, context);
+			params[index].progressive = progressive;
+			params[index].region = region;
+			params[index].context = context;
+			params[index].tile = tile;
+
+			if (!(work_objects[index] = CreateThreadpoolWork(
+			          progressive_process_tiles_tile_work_callback, (void*)&params[index],
+			          &progressive->rfx_context->priv->ThreadPoolEnv)))
+			{
+				WLog_ERR(TAG, "CreateThreadpoolWork failed.");
+				status = -1;
 				break;
+			}
 
-			case PROGRESSIVE_WBT_TILE_UPGRADE:
-				status = progressive_decompress_tile_upgrade(progressive, tile, region, context);
-				break;
-			default:
-				WLog_Print(progressive->log, WLOG_ERROR, "Invalid block type %04 (%s)" PRIx16,
-				           tile->blockType, progressive_get_block_type_string(tile->blockType));
-				return -42;
+			SubmitThreadpoolWork(work_objects[index]);
+			close_cnt = index + 1;
+		}
+		else
+		{
+			switch (tile->blockType)
```

Comment:
Maybe just call the thread callback here too to remove duplicate code?

Reply:
Done.